lua/orgmode/org/links/handlers/internal.lua

```diff
@@ -0,0 +1,114 @@
+local Org = require('orgmode')
+local utils = require('orgmode.utils')
+local Link = require('orgmode.org.links.link_handler')
+
+---@class OrgLinkHandlerInternal:OrgLinkHandler
+local Internal = Link:new()
+
+function Internal:new(protocol)
+  ---@class OrgLinkHandlerInternal
+  local this = Link:new(protocol)
+  setmetatable(this, self)
+  self.__index = self
+  return this
+end
+
+---@param target string | nil
+---@param disallow_file boolean?
+---@return OrgLinkHandlerInternal | OrgLinkHandlerFile | nil
+function Internal.parse(target, disallow_file)
```

Comment:
This whole class is doing a bit too much, and also has the checks that we are trying to avoid with the sources.
Each source should do this check and know how to parse the value from the target. Some example code:

```lua
local sources = { Headline, CustomId, LineNumber, File, Plain }

for _, source in ipairs(sources) do
  local parsed = source:parse(target)
  if parsed then
  	return parsed
  end
end
```

This should happen in the top level class, not here.

Reply:
I split this up into a separate handler because these are a bit of a special case, on account of not having a `<protocol>:` prefix.

It seemed redundant to me to repeat the `<protocol>:` prefix handling in each of the `LinkHandler`s individually, so I put this logic in `OrgLinks`, but these five cases are special in the fact that they don't have prefixes.

These are also special in that these are the only allowed targets for `id:` and `file:` links after the `::`, so it made handling of that easier.

`file:` is an even more special case because its `<protocol>:` prefix is optional, so if we say each 'internal' link should be able to parse itself in its entirety, either `File` needs special treatment to parse both the prefixed case and the non-prefixed case, or all `LinkHandler`s need to parse their own prefix.

It's a weird set of special cases, and I'm not sure there's a perfect solution, but I'm interested in your thoughts.